Always set imageId within DockerBuildImage on success (#818)

stevendpclark opened the pull request:
- Add a unit test that exposes the issue when the DockerBuildImage
   task is considered up to date an another task attempts to fetch
   the imageId value.
 - When the task is considered up-to-date, use the image id that was
   persisted to the file to set the value of imageId and return to
   other tasks to use it

closes #818

bmuschko commented on src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImageFunctionalTest.groovy:
```diff
@@ -270,6 +270,12 @@ class DockerBuildImageFunctionalTest extends AbstractGroovyDslFunctionalTest {
+                doLast {
+                    // Expose issues with getImageId returning null
+                    // https://github.com/bmuschko/gradle-docker-plugin/issues/818
+                    buildImage.getImageId().get()
```
We'll need to assert that the value is not null. We should also have this check in the different test cases at every step along the way.

stevendpclark replied:
@bmuschko Would you have any additional feedback on how I implemented this validation across the workflow in the review feedback commit?

bmuschko commented on src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy:
```diff
@@ -150,7 +150,9 @@ class DockerBuildImage extends AbstractDockerRemoteApiTask implements RegistryCr
-                    getDockerClient().inspectImageCmd(file.text).exec()
+                    def fileImageId = file.text
+                    getDockerClient().inspectImageCmd(fileImageId).exec()
+                    imageId.set(fileImageId)
```
The value of the `imageId` property should be set before making the call to Docker. In case Docker communication throws an exception, we already have the field populated.

stevendpclark replied:
Interesting, okay I performed the set afterwards in case the inspectImageCmd had thrown due to the image id not being found, setting it to an invalid value would have seemed incorrect in that use case. 

I hadn't thought about the use case that the communication to docker itself had failed. 

Is it worth making the distinction? I.E. set if the failure is due to a communication issue and assuming it's still a valid image id and not setting it if the image doesn't exist. A quick look through the docker-java seems to mention a specific NotFoundException (would need to validate this)

llaforest replied:
@bmuschko : I think we should set the imageIdt after the inspect fails because when it fails the task is set as not up-to-date and the imageId should not be set. It should be set only when the inspect succeeds in my opinion.
I will add the check in the tests though it is a good point.

@stevendpclark : At first I was catching only NotFoundException but then I realized that I could get more in other situations, miscommunication being one of them.

bmuschko replied:
@llaforest I think you are right here. Sorry, let's change this back.

bmuschko commented on src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImageFunctionalTest.groovy:
```diff
@@ -270,6 +270,12 @@ class DockerBuildImageFunctionalTest extends AbstractGroovyDslFunctionalTest {
+
+                doLast {
+                    // Expose issues with getImageId returning null
+                    // https://github.com/bmuschko/gradle-docker-plugin/issues/818
```
I propose you annotate the test method with `@Issue` and point to the relevant Github issue as value.

stevendpclark replied:
Ah neat, I didn't know about this annotation. Will do.

bmuschko commented on src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImageFunctionalTest.groovy:
```diff
@@ -533,6 +551,26 @@ class DockerBuildImageFunctionalTest extends AbstractGroovyDslFunctionalTest {
+
+    private static String imageIdValidation() {
+        """
+            tasks.withType(DockerBuildImage).all { Task task ->
```
The call to `all` is not required here. It will work without. Would you mind taking it off?

bmuschko commented on src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImageFunctionalTest.groovy:
```diff
@@ -533,6 +551,26 @@ class DockerBuildImageFunctionalTest extends AbstractGroovyDslFunctionalTest {
+    private static String imageIdValidation() {
+        """
+            tasks.withType(DockerBuildImage).all { Task task ->
+                def assertTask = tasks.create("assertImageIdFor\${task.name}"){
```
Would probably make sense to upper case the first letter of the task name so that the naming sticks to camel case notation.

bmuschko commented on src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImageFunctionalTest.groovy:
```diff
@@ -533,6 +551,26 @@ class DockerBuildImageFunctionalTest extends AbstractGroovyDslFunctionalTest {
+                    def dependantTask = task
+                    doLast {
+                        def value = dependantTask.getImageId().getOrNull()
+                        if(value == null || !(value ==~ /^\\w+\$/)) {
```
Just as a side note - not required to change. It's probably good enough to use `assert` instead of having to throw a `GradleException`.

stevendpclark commented:
Review feedback was implemented, I rebased my branch and squashed the commits into a single one. Please let me know if there is any additional modifications you would like me to make.

bmuschko commented on src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImageFunctionalTest.groovy:
```diff
@@ -533,6 +551,26 @@ class DockerBuildImageFunctionalTest extends AbstractGroovyDslFunctionalTest {
+    private static String imageIdValidation() {
+        """
+            tasks.withType(DockerBuildImage) { Task task ->
+                def assertTask = tasks.create("AssertImageIdFor\${task.name}"){
```
What I meant here was "upper case the value of the expression $task.name". Right now the task name starts with a upper case which is not standard convention. So this should look like `assertImageIdFor\${task.name.capitalize()}`. Can you please change that?

stevendpclark replied:
Yeah that makes more sense. Okay will tweak the commit now. 
Thanks for all the time for the reviews!

hugares commented:
Can this PR be included in v4.10.0 ?

stevendpclark commented:
Seems like the travis-ci build failed but for another functional test

> com.bmuschko.gradle.docker.tasks.container.DockerCreateContainerFunctionalTest > with autoRemove set, the container is removed after stopping FAILED
    org.spockframework.runtime.ConditionFailedWithExceptionError at DockerCreateContainerFunctionalTest.groovy:170
        Caused by: java.lang.NullPointerException at DockerCreateContainerFunctionalTest.groovy:170

Is this a known issue with a test running in the ci build system?

bmuschko commented:
@hugares Should be fine to include in 4.10.0.
@stevendpclark I have seen this test fail before. It's somewhat flaky. We'll need to investigate.
